File: sdks/php/src/ZivonPay.php
<?php

namespace ZivonPay;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\RequestException;

/**
 * ZivonPay PHP SDK
 * Official SDK for ZivonPay Payment Aggregator
 */
class ZivonPay
{
    private string $keyId;
    private string $keySecret;
    private string $baseUrl;
    private GuzzleClient $client;

    /**
     * Create a new ZivonPay client
     *
     * @param string $keyId API key ID
     * @param string $keySecret API secret key
     * @param string $environment 'sandbox' or 'production'
     * @param int $timeout Request timeout in seconds
     */
    public function __construct(
        string $keyId,
        string $keySecret,
        string $environment = 'sandbox',
        int $timeout = 30
    ) {
        $this->keyId = $keyId;
        $this->keySecret = $keySecret;

        // Determine base URL
        $this->baseUrl = $environment === 'production'
            ? 'https://api.zivonpay.com/v1'
            : 'https://sandbox.api.zivonpay.com/v1';

        // Configure HTTP client
        $this->client = new GuzzleClient([
            'base_uri' => $this->baseUrl,
            'timeout' => $timeout,
            'auth' => [$keyId, $keySecret],
            'headers' => [
                'Content-Type' => 'application/json',
                'User-Agent' => 'ZivonPay-PHP/1.0.0',
            ],
        ]);
    }

    /**
     * Create a new order
     *
     * @param array $data Order data
     * @param string|null $idempotencyKey Optional idempotency key
     * @return array Order data
     * @throws ZivonPayException
     */
    public function createOrder(array $data, ?string $idempotencyKey = null): array
    {
        $headers = [];
        if ($idempotencyKey !== null) {
            $headers['X-Idempotency-Key'] = $idempotencyKey;
        }

        return $this->request('POST', '/orders', $data, $headers);
    }

    /**
     * Fetch order by ID
     *
     * @param string $orderId Order UUID
     * @return array Order data
     * @throws ZivonPayException
     */
    public function fetchOrder(string $orderId): array
    {
        return $this->request('GET', "/orders/{$orderId}");
    }

    /**
     * List orders
     *
     * @param int $skip Number of records to skip
     * @param int $limit Number of records to return
     * @return array List response
     * @throws ZivonPayException
     */
    public function listOrders(int $skip = 0, int $limit = 10): array
    {
        return $this->request('GET', '/orders', null, [], [
            'skip' => $skip,
            'limit' => $limit,
        ]);
    }

    /**
     * Fetch payment by ID
     *
     * @param string $paymentId Payment UUID
     * @return array Payment data
     * @throws ZivonPayException
     */
    public function fetchPayment(string $paymentId): array
    {
        return $this->request('GET', "/payments/{$paymentId}");
    }

    /**
     * List payments
     *
     * @param int $skip Number of records to skip
     * @param int $limit Number of records to return
     * @return array List response
     * @throws ZivonPayException
     */
    public function listPayments(int $skip = 0, int $limit = 10): array
    {
        return $this->request('GET', '/payments', null, [], [
            'skip' => $skip,
            'limit' => $limit,
        ]);
    }

    /**
     * Verify webhook signature
     *
     * @param string $payload Raw request body
     * @param string $signature Signature from header
     * @param int $timestamp Timestamp from header
     * @param string $secret Webhook secret
     * @param int $tolerance Time tolerance in seconds
     * @return bool True if valid
     */
    public static function verifyWebhookSignature(
        string $payload,
        string $signature,
        int $timestamp,
        string $secret,
        int $tolerance = 300
    ): bool {
        // Check timestamp tolerance
        $currentTime = time();
        if (abs($currentTime - $timestamp) > $tolerance) {
            return false;
        }

        // Create expected signature
        $signedPayload = "{$timestamp}.{$payload}";
        $expectedSignature = hash_hmac('sha256', $signedPayload, $secret);

        // Compare signatures
        return hash_equals($signature, $expectedSignature);
    }

    /**
     * Make HTTP request
     *
     * @param string $method HTTP method
     * @param string $path API path
     * @param array|null $data Request body
     * @param array $headers Additional headers
     * @param array $params Query parameters
     * @return array Response data
     * @throws ZivonPayException
     */
    private function request(
        string $method,
        string $path,
        ?array $data = null,
        array $headers = [],
        array $params = []
    ): array {
        try {
            $options = ['headers' => $headers];

            if (!empty($data)) {
                $options['json'] = $data;
            }

            if (!empty($params)) {
                $options['query'] = $params;
            }

            $response = $this->client->request($method, $path, $options);
            $body = (string) $response->getBody();

            return json_decode($body, true);
        } catch (RequestException $e) {
            if ($e->hasResponse()) {
                $response = $e->getResponse();
                $body = (string) $response->getBody();
                $errorData = json_decode($body, true);

                if (isset($errorData['error'])) {
                    throw new ZivonPayException($errorData['error']);
                }
            }

            throw new ZivonPayException([
                'code' => 'REQUEST_ERROR',
                'description' => $e->getMessage(),
                'source' => 'sdk',
            ]);
        }
    }
}

/**
 * ZivonPay Exception
 */
class ZivonPayException extends \Exception
{
    public string $code;
    public string $description;
    public string $source;
    public ?string $step;
    public ?string $reason;
    public array $metadata;

    public function __construct(array $errorData)
    {
        $this->code = $errorData['code'] ?? 'UNKNOWN_ERROR';
        $this->description = $errorData['description'] ?? 'Unknown error';
        $this->source = $errorData['source'] ?? 'unknown';
        $this->step = $errorData['step'] ?? null;
        $this->reason = $errorData['reason'] ?? null;
        $this->metadata = $errorData['metadata'] ?? [];

        parent::__construct($this->description);
    }
}

File: sdks/go/zivonpay.go
// Package zivonpay provides the official Go SDK for ZivonPay Payment Aggregator
package zivonpay

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

const (
	sandboxBaseURL    = "https://sandbox.api.zivonpay.com/v1"
	productionBaseURL = "https://api.zivonpay.com/v1"
	defaultTimeout    = 30 * time.Second
)

// Client is the ZivonPay API client
type Client struct {
	keyID      string
	keySecret  string
	baseURL    string
	httpClient *http.Client
}

// Config holds the configuration for the client
type Config struct {
	KeyID       string
	KeySecret   string
	Environment string // "sandbox" or "production"
	Timeout     time.Duration
}

// NewClient creates a new ZivonPay client
func NewClient(config Config) *Client {
	baseURL := sandboxBaseURL
	if config.Environment == "production" {
		baseURL = productionBaseURL
	}

	timeout := config.Timeout
	if timeout == 0 {
		timeout = defaultTimeout
	}

	return &Client{
		keyID:     config.KeyID,
		keySecret: config.KeySecret,
		baseURL:   baseURL,
		httpClient: &http.Client{
			Timeout: timeout,
		},
	}
}

// CustomerInfo represents customer information
type CustomerInfo struct {
	Name   string `json:"name"`
	Mobile string `json:"mobile"`
	Email  string `json:"email,omitempty"`
}

// OrderCreateRequest represents order creation request
type OrderCreateRequest struct {
	Amount   int                    `json:"amount"`
	Currency string                 `json:"currency"`
	Receipt  string                 `json:"receipt"`
	Customer CustomerInfo           `json:"customer"`
	Notes    map[string]interface{} `json:"notes,omitempty"`
}

// Order represents an order
type Order struct {
	ID           string                 `json:"id"`
	Entity       string                 `json:"entity"`
	Amount       int                    `json:"amount"`
	Currency     string                 `json:"currency"`
	Status       string                 `json:"status"`
	Receipt      string                 `json:"receipt"`
	UPIIntentURL string                 `json:"upi_intent_url,omitempty"`
	QRCodeURL    string                 `json:"qr_code_url,omitempty"`
	Notes        map[string]interface{} `json:"notes"`
	CreatedAt    int64                  `json:"created_at"`
	ExpiresAt    int64                  `json:"expires_at,omitempty"`
	PaidAt       int64                  `json:"paid_at,omitempty"`
}

// Payment represents a payment
type Payment struct {
	ID               string `json:"id"`
	Entity           string `json:"entity"`
	OrderID          string `json:"order_id"`
	Amount           int    `json:"amount"`
	Currency         string `json:"currency"`
	Status           string `json:"status"`
	PayerVPA         string `json:"payer_vpa,omitempty"`
	RRN              string `json:"rrn,omitempty"`
	TransactionID    string `json:"transaction_id,omitempty"`
	BankName         string `json:"bank_name,omitempty"`
	ErrorCode        string `json:"error_code,omitempty"`
	ErrorDescription string `json:"error_description,omitempty"`
	CreatedAt        int64  `json:"created_at"`
	CapturedAt       int64  `json:"captured_at,omitempty"`
}

// ListResponse represents a list response
type ListResponse struct {
	Entity string      `json:"entity"`
	Count  int         `json:"count"`
	Data   interface{} `json:"data"`
}

// ErrorResponse represents an error from the API
type ErrorResponse struct {
	Code        string                 `json:"code"`
	Description string                 `json:"description"`
	Source      string                 `json:"source"`
	Step        string                 `json:"step,omitempty"`
	Reason      string                 `json:"reason,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// Error implements the error interface
func (e *ErrorResponse) Error() string {
	return fmt.Sprintf("%s: %s", e.Code, e.Description)
}

// CreateOrder creates a new order
func (c *Client) CreateOrder(req OrderCreateRequest, idempotencyKey string) (*Order, error) {
	headers := make(map[string]string)
	if idempotencyKey != "" {
		headers["X-Idempotency-Key"] = idempotencyKey
	}

	var order Order
	err := c.request("POST", "/orders", req, &order, headers)
	if err != nil {
		return nil, err
	}

	return &order, nil
}

// FetchOrder fetches an order by ID
func (c *Client) FetchOrder(orderID string) (*Order, error) {
	var order Order
	err := c.request("GET", fmt.Sprintf("/orders/%s", orderID), nil, &order, nil)
	if err != nil {
		return nil, err
	}

	return &order, nil
}

// ListOrders lists orders
func (c *Client) ListOrders(skip, limit int) (*ListResponse, error) {
	path := fmt.Sprintf("/orders?skip=%d&limit=%d", skip, limit)

	var response ListResponse
	err := c.request("GET", path, nil, &response, nil)
	if err != nil {
		return nil, err
	}

	return &response, nil
}

// FetchPayment fetches a payment by ID
func (c *Client) FetchPayment(paymentID string) (*Payment, error) {
	var payment Payment
	err := c.request("GET", fmt.Sprintf("/payments/%s", paymentID), nil, &payment, nil)
	if err != nil {
		return nil, err
	}

	return &payment, nil
}

// ListPayments lists payments
func (c *Client) ListPayments(skip, limit int) (*ListResponse, error) {
	path := fmt.Sprintf("/payments?skip=%d&limit=%d", skip, limit)

	var response ListResponse
	err := c.request("GET", path, nil, &response, nil)
	if err != nil {
		return nil, err
	}

	return &response, nil
}

// VerifyWebhookSignature verifies webhook signature
func VerifyWebhookSignature(payload, signature string, timestamp int64, secret string, tolerance int64) bool {
	// Check timestamp tolerance
	currentTime := time.Now().Unix()
	if abs(currentTime-timestamp) > tolerance {
		return false
	}

	// Create expected signature
	signedPayload := fmt.Sprintf("%d.%s", timestamp, payload)
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(signedPayload))
	expectedSignature := hex.EncodeToString(h.Sum(nil))

	// Compare signatures
	return hmac.Equal([]byte(signature), []byte(expectedSignature))
}

func (c *Client) request(method, path string, body interface{}, result interface{}, headers map[string]string) error {
	url := c.baseURL + path

	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return err
	}

	// Set basic auth
	req.SetBasicAuth(c.keyID, c.keySecret)

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "ZivonPay-Go/1.0.0")

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if resp.StatusCode >= 400 {
		var errorResp struct {
			Error ErrorResponse `json:"error"`
		}
		if err := json.Unmarshal(respBody, &errorResp); err != nil {
			return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBody))
		}
		return &errorResp.Error
	}

	if result != nil {
		if err := json.Unmarshal(respBody, result); err != nil {
			return err
		}
	}

	return nil
}

func abs(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}
